cli/CampaignConfigValidation/CampaignValidator.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation;
+
+use WMDE\Fundraising\Frontend\BucketTesting\CampaignCollection;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\CampaignValidationRuleInterface;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\DefaultBucketRule;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\ReferenceUrlRule;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\StartAndEndTimeRule;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\UniqueBucketRule;
+
+/**
+ * Validates campaign data by looking for inconsistencies and logical errors
+ * YAML validity is checked in during loading of campaign configuration
+ *
+ * @see \WMDE\Fundraising\Frontend\BucketTesting\CampaignConfiguration
+ * @license GNU GPL v2+
+ */
+class CampaignValidator {
+
+	private $campaignCollection;
+	private $hasValidated = false;
+	private $errors = [];
+
+	public function __construct( CampaignCollection $campaignCollection ) {
+		$this->campaignCollection = $campaignCollection;
+	}
+
+	public function isPassing(): bool {
+		return empty( $this->getErrors() );
+	}
+
+	public function getErrors(): array {
+		$this->validate();
+
+		return $this->errors;
+	}
+
+	private function validate() {
+		if ( $this->hasValidated ) {
+			return;
+		}
+
+		/**
+		 * @var $campaign \WMDE\Fundraising\Frontend\BucketTesting\Campaign
+		 */
+		foreach ( $this->campaignCollection->getIterator() as $campaign ) {
+			foreach ( $this->getRules() as $rule ) {
```

Comment:
Could you move rules iteration into a `validateCampaign` method? that way we can keep the nesting level low.

A small performance improvement over instantiating new rules for every campaign could be moving out the error tracking from the rules and into an `ErrorLogger` that gets passed to every rule.